pallets/author-filter/src/lib.rs

```diff
@@ -0,0 +1,119 @@
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Small pallet responsible determining which accounts are eligible to author at the current
+//! block height.
+//!
+//! Currently this pallet is tightly coupled to our stake pallet, but this design
+//! should be generalized in the future.
+//!
+//! Using a randomness beacon supplied by the `Randomness` trait, this pallet takes the set of
+//! currently staked accounts from pallet stake, and filters them down to a pseudorandom subset.
+//! The current technique gives no preference to any particular author. In the future, we could
+//! disfavor authors who are authoring a disproportionate amount of the time in an attempt to
+//! "even the playing field".
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet;
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::Randomness;
+	use frame_support::traits::Vec;
+	use frame_system::pallet_prelude::*;
+	use sp_core::H256;
+
+	/// The Author Filter pallet
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	// The maximum number of eligible authors at each hight.
+	// TODO make this part of the config trait. Or maybe express it as a percent.
+	const MAX_ELIGIBLE: usize = 3;
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + stake::Config {
+		/// The overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Deterministic on-chain pseudo-randomness used to do the filtering
+		type RandomnessSource: Randomness<H256>;
+	}
+
+	impl<T: Config> author_inherent::CanAuthor<T::AccountId> for Pallet<T> {
+		fn can_author(account: &T::AccountId) -> bool {
+			CurrentEligible::<T>::get().contains(account)
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		// At the beginning of each block, we calculate the set of eligible authors for this block.
+		// TODO it might make more sense to calculate for the next block at the end of this block.
+		fn on_initialize(_: T::BlockNumber) -> Weight {
+			//TODO only need to grab randomness in else clause.
+			// For now its here to support the debugging event
+			let randomness = T::RandomnessSource::random(&*b"author_filter");
+			let mut staked: Vec<T::AccountId> = stake::Module::<T>::validators();
+
+			// Reduce it to a subset if there are more staked then the max eligible
+			let eligible_subset = if staked.len() <= MAX_ELIGIBLE as usize {
+				staked.clone()
+			} else {
+				let mut eligible = Vec::new();
+				for i in 0..MAX_ELIGIBLE {
+					// Calculate the index by grabbing the corresponding byte out of the randomness
+					// This will only work when MAX_ELIGIBLE < 32 because that's how many bytes
+					// there are. There's a lot hacky about this POC.
+					let index = randomness.as_fixed_bytes()[i] as usize;
```

Comment:
is this index always different for the different `i` in `0..MAX_ELIGIBLE`? If so, this guarantees every selected will be a different account and my previous comment can be resolevd.